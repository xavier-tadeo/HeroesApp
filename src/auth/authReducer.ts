import { types } from "../types/types";

// const state = {
//   name: "arlet",
//   logged: true
// }

type TAction = {
  type: string;
  payload: {
    name: string;
    email: string;
  };
};

export const authReducer = (state = {}, action: TAction) => {
  switch (action.type) {
    case types.login:
      return {
        ...action.payload,
        logged: true,
      };

    case types.logout:
      return {
        logged: false,
      };

    default:
      return state;
  }
};
