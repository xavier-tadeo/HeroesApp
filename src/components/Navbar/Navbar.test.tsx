import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ReactTestRenderer from "react-test-renderer";
import { AuthContext } from "../../auth/authContext";
import { types } from "../../types/types";
import { Navbar } from "./Navbar";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Given a Navbar component", () => {
  const contextValue = {
    user: {
      name: "Pedro",
      logged: true,
    },
    dispatch: jest.fn(),
  };
  describe("When it render", () => {
    test("It should show correcty form", () => {
      const view = ReactTestRenderer.create(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/"]}>
            <Navbar />
          </MemoryRouter>
        </AuthContext.Provider>
      );

      expect(view.toJSON()).toMatchSnapshot();
    });

    test("It should show name Predo in the document", () => {
      render(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/"]}>
            <Navbar />
          </MemoryRouter>
        </AuthContext.Provider>
      );

      const expectName = screen.getByText("Pedro");

      expect(expectName).toBeInTheDocument();
    });
  });

  describe("When clicked logout button", () => {
    test("It should called the navigate and dispatch with arguments", () => {
      const action = {
        type: types.logout,
        payload: {
          name: "",
          email: "",
        },
      };

      render(
        <AuthContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={["/"]}>
            <Navbar />
          </MemoryRouter>
        </AuthContext.Provider>
      );

      const button = screen.getByRole("button");
      fireEvent.click(button);

      expect(contextValue.dispatch).toHaveBeenCalledWith(action);
      expect(mockNavigate).toHaveBeenCalledWith("/login", { replace: true });
    });
  });
});
