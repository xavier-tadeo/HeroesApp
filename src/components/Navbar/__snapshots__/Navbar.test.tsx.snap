// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Navbar component When it render It should show correcty form 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    Home
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        className="nav-item nav-link"
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        className="nav-item nav-link"
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        className="nav-item nav-link"
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
  >
    <ul
      className="navbar-nav ml-auto"
    >
      <span
        className="nav-item anv-link text-info"
      >
        Pedro
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
